cuML/src/randomforest/randomforest.cu

```diff
@@ -0,0 +1,308 @@
+int * rfClassifier<T>::predict(const T * input, int n_rows, int n_cols, bool verbose) const {
+
+	ASSERT(this->trees, "Cannot predict! No trees in the forest.");
+	ASSERT((n_rows > 0), "Invalid n_rows %d", n_rows);
+	ASSERT((n_cols > 0), "Invalid n_cols %d", n_cols);
+	int * preds = new int[n_rows];
+
+	int row_size = n_cols;
+
+	for (int row_id = 0; row_id < n_rows; row_id++) {
+
+		if (verbose) {
+			std::cout << "\n\n";
+			std::cout << "Predict for sample: ";
+			for (int i = 0; i < n_cols; i++) std::cout << input[row_id*row_size + i] << ", ";
+			std::cout << std::endl;
+		}
+
+		std::map<int, int> prediction_to_cnt;
+		std::pair<std::map<int, int>::iterator, bool> ret;
+		int max_cnt_so_far = 0;
+		int majority_prediction = -1;
+
+		for (int i = 0; i < this->n_trees; i++) {
+			//Return prediction for one sample.
+			if (verbose) {
+				std::cout << "Printing tree " << i << std::endl;
+				this->trees[i].print();
+			}
+			int * prediction = this->trees[i].predict(&input[row_id * row_size], 1, n_cols, verbose);
+			ret = prediction_to_cnt.insert(std::pair<int, int>(*prediction, 1));
+			delete [] prediction;
+			if (!(ret.second)) {
+				ret.first->second += 1;
+			}
+			if (max_cnt_so_far < ret.first->second) {
+				max_cnt_so_far = ret.first->second;
+				majority_prediction = ret.first->first;
+			}
+		}
+
+		preds[row_id] = majority_prediction;
+	}
+	return preds;
+}
+
+
+/**
+ * @brief Predict input data and validate against ref_labels.
+ * @tparam T: data type for input data (float or double).
+ * @param[in] input: test data (n_rows samples, n_cols features) in row major format. CPU pointer.
+ * @param[in] ref_labels: label values for cross validation (n_rows elements); CPU pointer.
+ * @param[in] n_rows: number of  data samples.
+ * @param[in] n_cols: number of features (excluding target feature).
+ * @param[in] verbose: flag for debugging purposes.
+ */
+template<typename T>
+RF_metrics rfClassifier<T>::cross_validate(const T * input, const int * ref_labels, int n_rows, int n_cols, bool verbose) {
+
+	int * predictions = predict(input, n_rows, n_cols, verbose);
+
+	unsigned long long correctly_predicted = 0ULL;
+	for (int i = 0; i < n_rows; i++) {
+		correctly_predicted += (predictions[i] == ref_labels[i]);
+	}
+
+	float accuracy = correctly_predicted * 1.0f/n_rows;
+	RF_metrics stats(accuracy);
+	if (verbose) stats.print();
+
+	delete [] predictions;
+
+	/* TODO: Potentially augment RF_metrics w/ more metrics (e.g., precision, F1, etc.).
+	   For non binary classification problems (i.e., one target and  > 2 labels), need avg for each of these metrics */
+	return stats;
+}
+
+template class rf<float>;
+template class rf<double>;
+
+template class rfClassifier<float>;
+template class rfClassifier<double>;
+
+// Stateless API functions: fit, predict and cross_validate
+
+void fit(rfClassifier<float> * rf_classifier, const cumlHandle& user_handle, float * input, int n_rows, int n_cols, int * labels, int n_unique_labels) {
```

Comment:
minor knitpick! Please always pass the `cumlHandle` as first argument to the fit method.